src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionDiagnosticAnalyzer.Analyzer.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertSwitchStatementToExpression
+{
+    internal sealed partial class ConvertSwitchStatementToExpressionDiagnosticAnalyzer
+    {
+        private sealed class Analyzer : OperationVisitor<object, AnalysisResult>
+        {
+            private Analyzer()
+            {
+            }
+
+            public static bool CanConvertToSwitchExpression(ISwitchOperation operation)
+            {
+                // When results are combined, "Neutral" can be overridden, but "Failure" can't.
+                // The result might remain neutral at the end, e.g. when all cases are `break;` alone,
+                // in which case we cannot convert this to a switch expression.
+                var analysisResult = new Analyzer().VisitSwitch(operation, unused: null);
+                return !analysisResult.IsNeutral && !analysisResult.IsFailure;
+            }
+
+            public override AnalysisResult VisitSwitch(ISwitchOperation operation, object unused)
+            {
+                // Fail if the switch statement is empty or any of sections have more than one `case` label
+                // Once we have "or" patterns, we can relax this to accept multi-case sections.
+                if (operation.Cases.Length == 0 || operation.Cases.Any(@case => @case.Clauses.Length != 1))
+                {
+                    return AnalysisResult.Failure;
+                }
+
+                // Iterate over all sections and match section bodies to
+                // see if they are all well-formed for a switch expression
+                return Aggregate(operation.Cases, (result, @case) => AnalysisResult.Match(result, VisitList(@case.Body)));
+            }
```

Comment:
one thing i didn't see in the pr was any sort of check that the switch had a default case.  did i miss it?  it feels like if it doesn't have a default case we can't convert it since the old behavior would have been to skip the switch, but the new behavior will be to throw an exception.